### PR TITLE
use createContext/useContext for scope/api instances

### PR DESCRIPTION
**Pull request by joreilly**

### CODE REVIEW

File: web/src/main/kotlin/App.kt [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
     val (people, setPeople) = useState(emptyList<Assignment>())
 
-    useEffect(dependencies = listOf()) {
-        scope.launch {
-            setPeople(api.fetchPeople().people)
+    useEffectWithCleanup(dependencies = listOf()) {
+        appState.mainScope.launch {
+            setPeople(appState.peopleInSpaceApi.fetchPeople().people)
         }
+        return@useEffectWithCleanup { appState.mainScope.cancel() }
```
Comment (aaronoe) [inline web/src/main/kotlin/App.kt:17]: I'd lift the scope out of the app state and instead have a scope for this component. Right now this clean up would cancel all coroutines launched in that scope.
I'd use the useMemo hook to create a scope for this functional component

File: web/build.gradle.kts [hunk 2]
```diff
@@ -25,5 +25,9 @@ dependencies {
 }
 
 
-kotlin.target.browser { }
-
+kotlin {
+    target {
+        useCommonJs()
```
Comment (aaronoe) [inline web/build.gradle.kts:30]: Why do we need this?

File: web/src/main/kotlin/Main.kt [hunk 3]
```diff
@@ -1,9 +1,25 @@
...
 import react.child
+import react.createContext
 import react.dom.render
 import kotlin.browser.document
 
+
+
+val AppStateContext = createContext<AppState>()
+
+object AppState {
```
Comment (aaronoe) [inline web/src/main/kotlin/Main.kt:14]: I'd almost call this app dependencies and use the context api to provide these

File: web/build.gradle.kts (hunk 2 above)
Comment (joreilly) [reply web/build.gradle.kts:30]: I got error "When accessing module declarations from UMD, they must be marked by both @JsModule and @JsNonModule" when I added context stuff.....https://github.com/bdudelsack/kotlin-fullstack-example recommended adding this

File: web/src/main/kotlin/App.kt [hunk 4]
```diff
@@ -1,20 +1,22 @@
...
 import react.*
 import react.dom.*
 import kotlinx.coroutines.*
 
 
+
 val App = functionalComponent<RProps> { _ ->
-    val scope = MainScope()
-    val api = PeopleInSpaceApi()
+    val mainScope = MainScope()
```
Comment (aaronoe) [inline web/src/main/kotlin/App.kt:9]: In my projects I moved the mainScope declaration either into the body of `useEffectWithCleanup` or used 
`val mainScope = useMemo({ MainScope() }, arrayOf())`. In JetPack compose the [same concept exists right here](https://developer.android.com/reference/kotlin/androidx/compose/package-summary#remember)

**Pull request merged**